doc/developer/design/20220511_replicated_compute_introspection.md

```diff
@@ -0,0 +1,360 @@
+respective target replicas before sending them.
+
+#### Cleanup on Replica Restart
+
+When a replica is restarted for whatever reason, we need to make sure that the
+persist shards they log their introspection data to are cleared. Otherwise,
+we might end up with stale introspection data that was written by a previous
+incarnation of the replica and never retracted because subsequent incarnations
+were not aware of the stale data's existence.
+
+Replicas should clear their introspection shards as soon as they receive a
+`CreateInstance` command, before they initialize the introspection dataflows.
+
+`persist` does not offer a convenient API to clear shards, but it should
+still provide all necessary means. To clear a shard, we can:
+
+  1. Make a snapshot at the current time.
+  1. Issue retraction updates for each record in the snapshot.
+
+Note that this way of clearing shards in-place will not work if an old
+incarnation of the replica is still around logging new data to the same shard,
+as the presence of multiple writers can lead to inconsistent/undefined shard
+contents. This issue will be resolved in the future by the use of storage
+renditions.
+
+#### Cleanup on Replica Drop
+
+When a replica receives a `DropInstance` command, it should drop all its
+introspection shards.
+
+### ADAPTER
+
+#### Replica-specific Introspection Configuration
+
+In the coordinator, the compute introspection/logging configuration is
+currently global to clusters (i.e. part of the `CreateComputeInstance`
+operation). This must be changed to make the introspection configuration
+replica-specific (part of `CreateComputeInstanceReplica`).
+
+When a new replica is created, the coordinator creates a new random persist
+shard ID (`ShardId::new()`) for each log variant and stores these shard IDs
+persistently in stash. It then passes the shard IDs as part of the per-replica
+introspection configuration to `Controller::add_replica_to_instance`, which
+takes care of translating to the appropriate `CreateInstance` compute command.
+
+When a replica is dropped, the coordinator removes its introspection shard IDs
+together with the rest of the replica state.
+
+On the SQL-side nothing needs to be changed, users will still be able to
+configure introspection (granularity and debug mode) on a per-cluster basis
+only.
+
+#### Introspection Tables
+
+When a new replica is created, the coordinator creates a new replica-specific
+introspection table for each log variant.
+
+Replica IDs are global between all clusters, so introspection tables can be
+uniquely named by log variant and replica ID:
+
+```text
+<mz-log-variant>_<replica-id>
+```
+
+... where `<mz-log-variant>` is the value of `BuiltinLog::name` for the
+respective log variant. For example, for log variant `Timely(Operates)` and
+replica `5` the introspection table name is `mz_dataflow_operators_5`.
+
+When a replica is dropped, the coordinator removes all associated introspection
+tables.
```

Comment:
This part is referring to the table entries in the coordinator's catalog, to the storage/persist collections. I assumed that removing tables from the catalog does not interfere with existing dataflows that read from the storage collections. This might be wrong.

It just occurred to me that we disallow deletion of user tables that are still referenced by a view. We should probably do the same for introspection tables, to not end up with invalid views. This means either we have to disallow dropping a replica as long as there are views on its introspection tables, or dropping a replica should not drop its introspection tables (and neither its storage collections). In the seconds case we'd need some other way to clean up the introspection tables and their storage collections once they have no dependencies anymore.

Reply:
I also realized that we shouldn't expose introspection data as tables but as sources instead. The dependency issue still remains though.